src/Microsoft.IdentityModel.Logging/LogHelper.cs

```diff
@@ -371,5 +418,50 @@ public static object MarkAsNonPII(object arg)
         {
             return new NonPII(arg);
         }
+
+        /// <summary>
+        /// Creates a <see cref="LogEntry"/> by using the provided event level, exception argument, string argument and arguments list.
+        /// </summary>
+        /// <param name="level"><see cref="EventLevel"/></param>
+        /// <param name="innerException"><see cref="Exception"/></param>
+        /// <param name="message">The log message.</param>
+        /// <param name="args">An object array that contains zero or more objects to format.</param>
+        private static LogEntry WriteEntry(EventLevel level, Exception innerException, string message, params object[] args)
+        {
+            if (string.IsNullOrEmpty(message))
+                return null;
+
+            if (innerException != null)
+            {
+                // if PII is turned off and 'innerException' is a System exception only display the exception type
+                if (!IdentityModelEventSource.ShowPII && !LogHelper.IsCustomException(innerException))
+                    message = string.Format(CultureInfo.InvariantCulture, "Message: {0}, InnerException: {1}", message, innerException.GetType());
+                else // otherwise it's safe to display the entire exception message
+                    message = string.Format(CultureInfo.InvariantCulture, "Message: {0}, InnerException: {1}", message, innerException.Message);
+            }
+
+            LogEntry entry = new LogEntry();
+
+            // Logs basic information (library version, DateTime, whether PII is ON/OFF) once before any log messages are written.
```

Comment:
Technically not correct as multiple threads might hit this code path in parallel and write a few duplicate messages. Not sure how important it is in this context. Using any sorts of locks on the hot path should be also avoided.
Consider moving HeaderWritten up to ln 448.

Reply:
Thank you for catching it. I'll modify it.

Reply:
Toggling _headerWritten value right after verifying header hasn't been written yet.